src/ghostwriter/models/v1/mapping.py

```diff
@@ -0,0 +1,180 @@
+"""Definition of the mapping from a top-level route to its substituted form,
+as well as any hooks that should be called before passing the input route
+to be rewritten.
+"""
+
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+from string import Template
+from typing import Annotated
+
+from pydantic import AfterValidator, BaseModel, BeforeValidator, Field
+
+from ... import hooks
+from ...exceptions import (
+    HookError,
+    HookNotFoundError,
+    MatchNotFoundError,
+    ResolutionError,
+)
+from ..substitution import Parameters
+
+
+def canonicalize_source_route(v: str) -> str:
+    """Force source route to have one leading and one trailing slash."""
+    return f"/{v.strip('/')}/"
+
+
+def resolve_hooks(
+    v: list[str] | list[Callable[[Parameters], Awaitable[None]]],
```

Comment:
I think the type definition here is fine, but for complex types aliasing may help readability:

`Hook: TypeAlias = Callable[[Parameters], Awaitable[None]]`

But can also see the benefit ok keeping it as it is for a reader to know what the expected method signature is without having to look for "Hook".